### ISSUE TITLE
Do pre-bundle-add and pre-os-install disk space checks

### ISSUE THREAD

**Issue by matthewrsj** [clearlinux/swupd-client#266]
Instead of trying to do a bundle-add or OS installation and failing due to out-of-disk errors, perform a disk-space check before attempting the operation.

This can be coupled with the bundle-list --deps functionality and the fixed contentsize in the manifests to get a required disk space metric.

**Comment by icarus-sparry** [comment]
Of course this problem is not perfectly solvable, as there is a window between the disk space check and the install during which the space could be used. It gets even more complicated if you have multiple mounts. That doesn't mean to say that it is not worth attempting, just that it is not trivial.

**Comment by matthewrsj** [comment]
We could at least make an educated guess that the install will fail, no? Even the contentsize calculation will only be a ceiling since there can be overlapping files.

**Comment by matthewrsj** [comment]
@icarus-sparry do you have time to look at this? Even a best guess is better than what we have currently, which is just an avalanche of `not fixed` errors.

**Comment by icarus-sparry** [comment]
OK, will take a stab at it.

**Comment by matthewrsj** [comment]
This can potentially be made easier with the functionality being introduced in #399

**Comment by phmccarty** [comment]
Also worth mentioning here is that swupd-client should report `ENOSPC` errors correctly. Out-of-disk conditions are often inferred through other errors that are less common to see (e.g. verify failing to "fix" files and reporting `not fixed` repeatedly, when it would be nicer to see `failed to fix file: out of disk space` instead).

Should we track these reporting problems in a separate issue (or issues)?

**Comment by evil-brain** [comment]
IMHO reporting problems with out of disk space should be the first step toward resolving this issue.

When I started I saw this issue not just as a "check if we have enough space" problem, but more of a "swupd has problems when the FS is full" issue.

I've started by trying to fix the reporting and handling the out of space case in a sane manner.

Preemptive warning of space is good to have as well, but it makes more sense to me to start with handling the current errors/report them correctly.

With the potential for random mountpoints on the FS (along with the FS being used during updates), preemptive checking will always be a kind of 'best guess'. Unless we should check available space for each file IO operation... which seems like it'd be too much overhead.

A new issue might make sense, because there are a few places a full FS could cause problems. (ex when d/ling manifests after invoking search, d/ling bundles, decompressing, copying files...)
